File: packages/earthsci_toolkit/src/earthsci_toolkit/cli/determinism_adapter.py
"""Python determinism-conformance adapter (``CONFORMANCE_SPEC.md`` §5.5.4).

The thin bridge the cross-binding determinism harness
(``scripts/run-determinism-conformance.py``) invokes to exercise the **Python**
relational engine (:mod:`earthsci_toolkit.relational`) over the shared golden
fixtures in ``tests/conformance/determinism/manifest.json``. The runner discovers
it via ``$EARTHSCI_DETERMINISM_ADAPTER_PYTHON`` or as
``earthsci-determinism-adapter-python`` on ``PATH`` and calls::

    earthsci-determinism-adapter-python --manifest <manifest.json> --output <result.json>

For each fixture it runs the real producers over ``inputs.canonical`` and writes
the canonical index set, its byte-form serialization, and the dense-ID array in
Python's **native 0-based** emission base (the harness normalises via
``rank_base_pin``). Keep this thin — the contract lives in the engine, not here.
"""

from __future__ import annotations

import argparse
import json
import operator
import sys
from pathlib import Path
from typing import Any, Dict, List, Tuple

from earthsci_toolkit.relational import (
    canonical_index_set_json,
    distinct,
    group_aggregate,
    rank,
    serialize_canonical,
    skolem,
    skolem_edge,
)


def _directed_edges_from_faces(faces: List[List[Any]]) -> List[Tuple[Any, Any]]:
    """Traverse consecutive vertices of each face (with wraparound) into directed
    edges — the realistic producer step a mesh FAQ performs before skolem
    canonicalisation (mirrors the runner's reference shaping)."""
    edges: List[Tuple[Any, Any]] = []
    for face in faces:
        n = len(face)
        for i in range(n):
            edges.append((face[i], face[(i + 1) % n]))
    return edges


def _compute_fixture(fixture: Dict[str, Any]) -> Dict[str, Any]:
    primitive = fixture["primitive"]
    payload = fixture["inputs"]["canonical"]

    if primitive == "skolem_distinct_rank":
        mode = fixture.get("skolem")
        if "faces" in payload:
            edges = _directed_edges_from_faces(payload["faces"])
        elif "tuples" in payload:
            edges = [tuple(t) for t in payload["tuples"]]
        else:
            raise ValueError(
                f"fixture {fixture['id']}: input needs 'faces' or 'tuples'"
            )

        if mode == "undirected":
            keys = [skolem_edge(a, b) for (a, b) in edges]
        elif mode == "directed":
            keys = [skolem(t) for t in edges]
        else:
            raise ValueError(f"fixture {fixture['id']}: unknown skolem mode {mode!r}")

        index_set = distinct(keys)
        serialized = canonical_index_set_json(keys)
        ranking = rank(keys)  # native 0-based
        dense = [ranking.ids[t] for t in index_set]

    elif primitive == "group_by_sum":
        pairs = group_aggregate(
            payload["rows"],
            key=lambda row: row[0],
            value=lambda row: row[1],
            op=operator.add,
        )
        index_set = pairs
        serialized = serialize_canonical(pairs)
        dense = list(range(len(pairs)))  # positions of the sorted distinct keys

    else:
        raise ValueError(f"fixture {fixture['id']}: unknown primitive {primitive!r}")

    return {
        "index_set": [list(row) for row in index_set],
        "serialized": serialized,
        "dense_ids_canonical": dense,
    }


def main(argv: "List[str] | None" = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--manifest", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args(argv if argv is not None else sys.argv[1:])

    with args.manifest.open() as f:
        manifest = json.load(f)

    fixtures: Dict[str, Any] = {}
    for fixture in manifest["fixtures"]:
        fixtures[fixture["id"]] = _compute_fixture(fixture)

    result = {"binding": "python", "fixtures": fixtures}
    args.output.parent.mkdir(parents=True, exist_ok=True)
    with args.output.open("w") as f:
        json.dump(result, f)
        f.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: packages/EarthSciSerialization.jl/scripts/determinism_adapter.jl
#!/usr/bin/env julia
# Julia determinism-conformance adapter (CONFORMANCE_SPEC.md §5.5.4). The thin
# bridge the cross-binding determinism harness (scripts/run-determinism-conformance.py)
# invokes to exercise the REAL Julia value-invention primitives
# (EarthSciSerialization.Relational: skolem / distinct / rank + group-by) over the
# shared golden fixtures in tests/conformance/determinism/manifest.json. The runner
# discovers it via $EARTHSCI_DETERMINISM_ADAPTER_JULIA or as
# earthsci-determinism-adapter-julia on PATH, and calls:
#
#     <adapter> --manifest <manifest.json> --output <result.json>
#
# For each fixture it runs the primitives over inputs.canonical AND every
# adversarial inputs.variants payload, writing the canonical serialized index set,
# its index_set, and the dense-ID array in Julia's NATIVE 1-based emission base
# (the runner normalizes via rank_base_pin: julia=1). Keep this thin — the contract
# lives in src/relational.jl, not here.

using EarthSciSerialization
const R = EarthSciSerialization.Relational
import JSON3

function parse_args(argv)
    manifest = nothing
    output = nothing
    i = 1
    while i <= length(argv)
        a = argv[i]
        if a == "--manifest"
            manifest = argv[i+1]; i += 2
        elseif a == "--output"
            output = argv[i+1]; i += 2
        else
            error("determinism_adapter: unknown argument $(repr(a))")
        end
    end
    (manifest === nothing || output === nothing) &&
        error("determinism_adapter: --manifest and --output are required")
    return manifest, output
end

# Recursively convert JSON3 values to native Dict/Vector so the relational
# primitives see plain Julia Int/String/Bool (not lazy JSON3 views).
to_native(x::JSON3.Object) = Dict{String,Any}(string(k) => to_native(v) for (k, v) in x)
to_native(x::JSON3.Array) = Any[to_native(v) for v in x]
to_native(x) = x

# Directed component lists for one payload: faces (consecutive vertices with
# wraparound) or pre-built tuples — mirrors the runner's reference shaping.
function edges_from_payload(payload)
    if haskey(payload, "faces")
        edges = Vector{Any}[]
        for face in payload["faces"]
            n = length(face)
            for i in 1:n
                push!(edges, Any[face[i], face[mod1(i + 1, n)]])
            end
        end
        return edges
    elseif haskey(payload, "tuples")
        return Vector{Any}[collect(Any, t) for t in payload["tuples"]]
    else
        error("payload needs 'faces' or 'tuples'")
    end
end

function compute_payload(fx, payload)
    primitive = fx["primitive"]
    if primitive == "skolem_distinct_rank"
        mode = fx["skolem"]
        edges = edges_from_payload(payload)
        keys = if mode == "undirected"
            [R.skolem_edge(e[1], e[2]) for e in edges]
        elseif mode == "directed"
            [R.skolem(Tuple(e)) for e in edges]
        else
            error("unknown skolem mode $(repr(mode))")
        end
        index_set = R.distinct(keys)
        serialized = R.canonical_index_set_json(keys)
        ranking = R.rank(keys)                       # native 1-based
        dense = [ranking.id[t] for t in index_set]
        return Dict{String,Any}(
            "index_set" => [collect(t) for t in index_set],
            "serialized" => serialized,
            "dense_ids_canonical" => dense,
        )
    elseif primitive == "group_by_sum"
        rows = payload["rows"]
        pairs = R.group_aggregate(rows; key = r -> r[1], value = r -> r[2], op = +)
        # Serialize the (key,sum) pairs in canonical form. Sums here are exact
        # integers, so reusing the index-set serializer over (k,v) tuples is
        # byte-correct (and keys are unique ⇒ no reorder/dedup).
        serialized = R.canonical_index_set_json([(first(p), last(p)) for p in pairs])
        return Dict{String,Any}(
            "index_set" => [[first(p), last(p)] for p in pairs],
            "serialized" => serialized,
            "dense_ids_canonical" => collect(1:length(pairs)),   # native 1-based
        )
    else
        error("unknown primitive $(repr(primitive))")
    end
end

function compute_fixture(fx)
    record = compute_payload(fx, fx["inputs"]["canonical"])
    variants = get(fx["inputs"], "variants", Dict{String,Any}())
    if !isempty(variants)
        record["variants"] = Dict{String,Any}(
            vname => compute_payload(fx, vpayload) for (vname, vpayload) in variants
        )
    end
    return record
end

function main(argv)
    manifest_path, output_path = parse_args(argv)
    manifest = to_native(JSON3.read(read(manifest_path, String)))

    fixtures = Dict{String,Any}()
    for fx in manifest["fixtures"]
        fixtures[fx["id"]] = compute_fixture(fx)
    end

    result = Dict{String,Any}("binding" => "julia", "fixtures" => fixtures)
    mkpath(dirname(abspath(output_path)))
    open(output_path, "w") do io
        JSON3.write(io, result)
        write(io, "\n")
    end
    return 0
end

exit(main(ARGS))
